--- MusicInfoSCR.pm	2006-01-28 18:19:17.000000000 +0100
+++ MusicInfoSCR.pm.new	2006-01-28 18:40:04.000000000 +0100
@@ -45,6 +45,10 @@
 use vars qw($VERSION);
 $VERSION = substr( q$Revision$, 10 );
 
+# Start plugin patch
+my %customitems = ();
+# End plugin patch
+
 sub getDisplayName() {
 	return 'PLUGIN_SCREENSAVER_MUSICINFO';
 }
@@ -268,6 +272,22 @@
 	$formatStrings{'(X_OF_Y)'} = '(X_OF_Y)';
 	$formatStrings{'PROGRESSBAR'} = 'PROGRESSBAR';
 
+	# Start plugin patch
+	no strict 'refs';
+	my @plugins = Slim::Buttons::Plugins::enabledPlugins();
+	for my $plugin (@plugins) {
+		if(UNIVERSAL::can("Plugins::$plugin","getMusicInfoSCRCustomItems") && UNIVERSAL::can("Plugins::$plugin","getMusicInfoSCRCustomItem")) {
+			$::d_plugins && Slim::Utils::Misc::msg("MusicInfoSCR: Getting items for: $plugin\n");
+			my $items = eval { &{"Plugins::${plugin}::getMusicInfoSCRCustomItems"}() };
+			for my $item (keys %$items) {
+				$customitems{$item} = "Plugins::${plugin}::getMusicInfoSCRCustomItem";
+				$formatStrings{$item} = $items->{$item};
+			}
+		}
+	}
+	use strict 'refs';
+	# End plugin patch
+
 	my %setupGroup = (
 		PrefOrder => [
 			'plugin_musicinfo_lineA', 'plugin_musicinfo_overlayA',
@@ -352,6 +372,10 @@
 sub setScreensaverMusicInfoMode() {
 	my $client = shift;
 	
+	# Start plugin patch
+	clearCustomItemCache($client);
+	# End plugin patch
+	
 	# initialize settings if they don't exist yet
 	if (not ($client->prefGet('plugin_musicinfo_lineA') || $client->prefGet('plugin_musicinfo_overlayA') ||
 			$client->prefGet('plugin_musicinfo_lineB') || $client->prefGet('plugin_musicinfo_overlayB') ||
@@ -556,6 +580,10 @@
 			$formattedString = $titleOnly;
 		}
 		
+		# Start plugin patch
+		$formattedString = getCustomItem($client,$formattedString);
+		# End plugin patch
+			
 		# if ARTIST/ALBUM etc. is empty, replace them by some "Now playing..."
 		if ($albumOrArtist) {
 			my $tmpFormattedString = $formattedString;
@@ -654,6 +682,37 @@
 	return ($displayCache{$client}->{'icons'}, $line . ("\x00" x ($displayCache{$client}->{'width'})));
 }
 
+# Start plugin patch
+sub getCustomItem()
+{
+	my $client = shift;
+	my $formattedString = shift;
+
+	no strict 'refs';
+
+	for my $item (keys %customitems) {
+		my $method = %customitems->{$item};
+		if ($formattedString =~ /$item/) {
+			$::d_plugins && Slim::Utils::Misc::msg("MusicInfoSCR ($client): Getting custom info for $item from: $method\n");
+			$formattedString = eval { &{"${method}"}($client,$formattedString) };
+		}
+	}
+	return $formattedString;
+}
+sub clearCustomItemCache()
+{
+	my $client = shift;
+
+	no strict 'refs';
+
+	for my $item (keys %customitems) {
+		$displayCache{$client}->{$item} = {
+			'songkey' => '',
+			'formattedString' => ''
+		};
+	}
+}
+# End plugin patch
 
 sub strings { return q^
 PLUGIN_SCREENSAVER_MUSICINFO
